Fix bug of using symbolic link dir as storage path

kangpinghuang opened the pull request:
There is a bug to use symbolic link directory as storage root path.
It is a problem that whether the path is canonical

kangpinghuang committed:
Fix bug of #307

There is a bug to use symbolic link directory as storage root path.
It is a problem that whether the path is canonical

imay commented on be/src/http/download_action.cpp:
```diff
@@ -254,7 +258,8 @@ Status DownloadAction::check_path_is_allowed(const std::string& file_path) {
     DCHECK_EQ(_download_type, ERROR_LOG);
-    if (FileSystemUtil::contain_path(_error_log_root_dir, file_path)) {
+    std::string canonical_file_path = canonical(file_path).string();
+    if (FileSystemUtil::contain_path(canonical(_error_log_root_dir).string(), canonical_file_path)) {
```
`_error_log_root_dir` should be set after canonical when initializing.

morningman replied:
why not just call canonical in DownloadAction construct method?

imay commented on be/src/http/download_action.cpp:
```diff
@@ -243,8 +246,9 @@ Status DownloadAction::check_token(HttpRequest *req) {
+    std::string canonical_file_path = canonical(file_path).string();
     for (auto& allow_path : _allow_paths) {
-        if (FileSystemUtil::contain_path(allow_path, file_path)) {
+        if (FileSystemUtil::contain_path(canonical(allow_path).string(), canonical_file_path)) {
```
same with `_error_log_root_dir ` to `_allow_paths`

imay commented:
@kangpinghuang 
And you need to change your commit log to describe what you fixed

kangpinghuang committed:
Fix bug of using symbolic link dir as storage path

In DownloadAction, checking fails by comparing canonical path with noncanonical path.
So fix the bug by convert all path to canonical path before comparison

imay commented on be/src/http/download_action.cpp:
```diff
@@ -47,16 +50,16 @@ const std::string TOKEN_PARAMETER = "token";
-
+    _download_type(NORMAL) {
+    for (auto& dir : allow_dirs) {
+        _allow_paths.emplace_back(std::move(canonical(dir).string()));
```
`canonical(dir).string()` is already a rvalue, you needn't to call `std::move`

kangpinghuang committed:
Delete unnecessary std::move

morningman commented:
ISSUE #307

imay reviewed (approved):
LGTM
